packages/expo-notifications/src/utils/updatePushTokenAsync.ts

```diff
@@ -0,0 +1,138 @@
+import { CodedError, Platform } from '@unimodules/core';
+import * as Application from 'expo-application';
+
+import { Registration } from '../DevicePushTokenAutoRegistration.fx';
+import ServerRegistrationModule from '../ServerRegistrationModule';
+import { DevicePushToken } from '../Tokens.types';
+import generateRetries from './generateRetries';
+import makeInterruptible from './makeInterruptible';
+
+export const [
+  updatePushTokenAsync,
+  hasPushTokenBeenUpdated,
+  interruptPushTokenUpdates,
+] = makeInterruptible<[DevicePushToken]>(updatePushTokenAsyncGenerator);
+
+async function* updatePushTokenAsyncGenerator(token: DevicePushToken) {
+  // Fetch the latest registration info from the persisted storage
+  const lastRegistrationInfo = yield ServerRegistrationModule.getLastRegistrationInfoAsync?.();
+  // If there is none, do not do anything.
+  if (!lastRegistrationInfo) {
+    return;
+  }
+
+  // Prepare request body
+  const lastRegistration: Registration = JSON.parse(lastRegistrationInfo);
+  // Persist `pendingDevicePushToken` in case the app gets killed
+  // before we finish registering to server.
+  await ServerRegistrationModule.setLastRegistrationInfoAsync?.(
+    JSON.stringify({
+      ...lastRegistration,
+      pendingDevicePushToken: token,
+    })
+  );
+
+  const body = {
+    ...lastRegistration.body,
+    // Information whether a token is applicable
+    // to development or production notification service
+    // should never be persisted as it can change between
+    // Xcode development and TestFlight/AppStore without
+    // backing store being resetted (development registration
+    // remains in production environment).
+    development: await shouldUseDevelopmentNotificationService(),
+    deviceToken: token.data,
+    type: getTypeOfToken(token),
+  };
+
+  const retriesIterator = generateRetries(async retry => {
+    try {
+      const response = await fetch(lastRegistration.url, {
+        method: 'POST',
+        headers: {
+          'content-type': 'application/json',
+        },
+        body: JSON.stringify(body),
+      }).catch(error => {
+        throw new CodedError(
+          'ERR_NOTIFICATIONS_NETWORK_ERROR',
+          `Error encountered while updating device push token in server: ${error}.`
+        );
+      });
```

Comment:
For example, If the `JSON.stringify` fails, we will throw a network error. I'm not sure if what we want here. Theoretically, the app is going to retry the call in that case, but the same error will occur. So in that way, the app might end up in the infinite retries loop. However, I believe that the other errors (than network one) are pretty unlikely here. So maybe we can leave it as it is.

But I lean toward that we should check if the network error occurred.

Reply:
Current implementation wraps as little code as possible with `try`/`catch`. According to [documentation](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/JSON/stringify#Exceptions), `JSON.stringify` throws:
- Throws a TypeError ("cyclic object value") exception when a circular reference is found.
- Throws a TypeError ("BigInt value can't be serialized in JSON") when trying to stringify a BigInt value.

None of those exceptions can happen with current implementation, so I believe it should be safe to retry always unless error is an `AbortError` (as current implementation does).